### ISSUE TITLE
[css-functions] enable the use of counter() inside calc()

### ISSUE THREAD

**Issue by psylok** [w3c/csswg-drafts#1026]
Feature request.
It would be nice to be able to use the counter() function inside of calc() function.
That would enable new possibilities on layouts.

I copy here the link to a thread which proposed it last august.
https://lists.w3.org/Archives/Public/www-style/2016Aug/0073.html

**Comment by AmeliaBR** [comment]
[Relevant section of the spec](https://drafts.csswg.org/css-lists-3/#counter-functions), which includes an inline issue discussing possible use cases.  But getting an issue on GitHub will hopefully provide an easier forum for other people to mention their use cases and/or possible implementation issues.

**Comment by MelSumner** [comment]
Hi, ran into a case today where this would have helped, so posting a comment here. 

I'm setting up some CSS for an image/logo slider. While I know right now that there are 8 images, I don't know how many there will be in the future (nor do I wish to have to know).  I'm using only CSS to animate it, and it would be useful to be able to say "for each one of the items that exist (, increment this property by x." 
I ended up using sass to do something like this: 
```    
@for $i from 1 through 8 {  
    &:nth-of-type(#{$i}) {
        $delay: calc(5s * #{$i});
	animation: slideIn 45s linear $delay infinite;
    }
}

**Comment by Loirooriol** [comment]
The problem with this proposal is that `counter()` returns a `<string>`, but `calc()` does not work with strings. For example, `counter(id, upper-latin)` might be `'A'`. How exactly is `calc()` supposed to know that this means `1`?

So I think we need some way to get the numeric value of a counter, either by adding some parameter to `counter()` or a new function. This should be allowed to appear anywhere an integer is expected, including (but not necessarily in) `calc()`. I wrote my thoughts in #1871.

**Comment by Crissov** [comment]
JFTR, I would hate to see `int()` or `str2num()` in CSS.

**Comment by tabatkins** [comment]
Yeah, an explicit parsing function is probably bad. No real need for it. But having something that can retrieve a counter value as a number (rather than going to the extra effort of formatting it as a string) would work.  (Same as how `attr()` has functionality to parse the attr value as a number or dimension.)

**Comment by upsuper** [comment]
Wouldn't allowing `counter()` in `calc()` either ruin any attempt to parallelize styling, or makes us need to delay calculating of every numeric value to used-value time?

**Comment by Loirooriol** [comment]
@upsuper Fair point, counter values are inherited from the immediately preceding element in document order instead of from the parent element as usual. So maybe it would be better to add a `sibling-index()` function as Tab Atkins proposed in https://github.com/w3c/csswg-drafts/issues/1869#issuecomment-340078522; it would be less powerful but I think it would cover most usecases.

However, note that the CSS Lists draft currently allows [`counter()`](https://drafts.csswg.org/css-lists/#counter-functions) anywhere a `<string>` is expected, doesn't this have the same problem? If it's possible to use a counter value as a string, it should also be possible to use it as an integer.

**Comment by upsuper** [comment]
> However, note that the CSS Lists draft currently allows counter() anywhere a <string> is expected, doesn't this have the same problem? If it's possible to use a counter value as a string, it should also be possible to use it as an integer.

Firstly, there are a lot fewer places where `<string>` is accepted, and they are usually relatively less complicated than `<length>` and its friends when it comes to layout, so expanding them in used-value time (like what is currently done for `content` property) may not be as bad.

Also, allowing `counter()` anywhere `<string>` is allowed is something new. [`counter()` functions are listed as independent item of `content` property in CSS2.](https://www.w3.org/TR/CSS2/generate.html#content) And I'm not aware of any browser who has implemented that anywhere outside `content`. That means, its feasibility may also be questioned, and my argument above can potentially be apply to that as well.

**Comment by upsuper** [comment]
> Firstly, there are a lot fewer places where `<string>` is accepted, and they are usually relatively less complicated than `<length>` and its friends when it comes to layout, so expanding them in used-value time (like what is currently done for content property) may not be as bad.

It may not be as bad, but it could still be very bad. One issue comes to my mind is that `<family-name>` can be a `<string>`, and `font-family` is inherited, so you may really want to make `counter()` be expanded in computed-value time (rather than in used-value time like what we do now for `content`), otherwise it can lead to some funny behavior. This is some case which seems to be neither useful nor easy to implement, which would be a native consequence if we allow `counter()` be in anywhere `<string>` is allowed.

I guess we should start this topic in a separate issue, now.

**Comment by kizu** [comment]
I would really, really love to have counter's value available inside `calc()`. The syntax I'd propose could be something like `counter-value()` that would return the counter's value as an integer. Other than the use-cases for `sibling-index`, this could be used for a lot more cases (not only [experimental](http://kizu.ru/en/blog/counters-and-stones/) ones). While I understand that that can be non-trivial to implement, the possibilities it would unlock would be tremendous.

**Comment by LeaVerou** [comment]
Two more use cases:

- Staggered animations by using it in `animation-delay`
- Reversing the z-index order by using it in `z-index`

Regarding syntax, I like @tabatkins's proposal for a new formatting argument, if an explicit syntax is required.
Although IMO ideally it would be nice if we could auto cast number-like strings to numbers in `calc()` akin to how many programming languages do it (including JS). It's not like strings actually do anything in `calc()`, so there's no disambiguation problem.

**Comment by jonjohnjohnson** [comment]
My sad excuse of a utility without this feature...

```css
...
.count--3 { --count: 3 }
.count--4 { --count: 4 }
.count--5 { --count: 5 }
...
.count > *:nth-child(1) { --count-current: 1 }
.count > *:nth-child(2) { --count-current: 2 }
.count > *:nth-child(3) { --count-current: 3 }
...
.el {
  animation-delay: calc((var(--count, 0) - var(--count-current, 0)) * 0.1s);
} 
```

**Comment by faceless2** [comment]
I like this in general and I'd love to see it implemented too. But I'd really, really hate implementing it.

All of the paged media layout engines have the same problem with counters, in particular "page" and "pages", although technically it's any counters incremented in the @page margin areas (still a largely theoretical constraint at the time of writing). The moment you hit one of these you have to consider a second layout pass - it's not always required, but for something like `span::after { content: counter(pages upper-roman); }` you really haven't got much choice.

You need the first layout pass to count the pages, or to work out with certainty which page your element is on (that's more of an issue with `target-counter(nnn, page)` to be fair). But once you start introducing counters to other properties - say `margin-top: calc(counter(pages) * 20px)` - you've introduced a loop: layout depends on values computed from an earlier pass of the layout, and so on.

It's not quite as awful for the "page" counter, or any counter other than "pages". But it's still a little complex: `orphans: calc(counter(page) * 20)` may force a page-break, which would change the value of orphans...

So although I'm not exactly against this, I just wanted to flag that it is almost unfeasibly hairy for some cases. Think  of the multiple passes required to stabilise layout for ::first-line, except it's a _whole document_ that needs to be stabilised. These would need to be considered if this goes anywhere.

**Comment by LeaVerou** [comment]
Won’t we get this for free once we have a function to convert between types, the likes of which has been discussed multiple times?

**Comment by steren** [comment]
I landed here because I assumed I would be able to do something like this in CSS:
```
li:nth-child(n) {
  background-color: hsl(calc(120 *n), 100%, 70%);
}
```

**Comment by yinonov** [comment]
working on a calendar like grid, was sure i'd be able to layout overlapping scheduled events by the use of counter-value. js got tons of practices for a conditional index. can't css have 1?

**Comment by ShahriarKh** [comment]
Another use case here:
```css
.item:nth-of-type(n) {
  transform: rotate(calc(n*10deg)); /* 10 comes from 360/totalItemsCount */
}
```

**Comment by Loirooriol** [comment]
@ShahriarKh Your use case seems better covered by #4559:
```css
.item { transform: rotate(calc(1turn * sibling-index() / sibling-count())) }
```
With a counter-based approach you would also need #3667 to know the totalItemsCount.

**Comment by ShahriarKh** [comment]
@LeaVerou 
> Could you elaborate a bit on this? `n` cannot be used in `calc()`, and it's unclear ....

Yeah I know this isn't possible. I find this discussion when I was playing with CSS: https://codepen.io/shahriarkh/pen/rNGwQMQ
I wanted to place some bars inside the circle like this: 
<img src="https://user-images.githubusercontent.com/31452340/146781743-49a59a88-4137-4e47-8c0e-7d1e98b992b5.png" alt="drawing" width="200px" />
Declaring rotation for every bar isn't a good idea, so I wondered it will be useful if we had something like this:
```css
.bar:nth-of-type(n) {  
    transform: rotate(calc(n * 36deg))
}
```

**Comment by TheophileDesmedt** [comment]
Here's another way this feature could be used : decreasing the size of siblings.

```css
.circles { 
  counter-reset: n;
 }
 
.circles circle {
counter-increment: n;
width:calc(100% / counter(n)  );
}
```

Is there any hope to get something like this in CSS one day or is that a task more suited for javascript ?

**Comment by JJanz** [comment]
And I got here to sadly find out this is still open and with yet another use case (I couldn't find it by scanning the ones here): counting DOM depth, which I intented use to indent only one element in nested tags.
Like:
```
body {
    counter-reset: tag-depth;
}
* > .indented {
    counter-increment: tag-depth;
}
.indented {
    margin-left: calc(1.5rem * counter-value(tag-depth);
}
```
I would totally solve this with recursive adding 1 to a custom property (like `* > .indented { --tag-depth: calc(var(--tag-depth) + 1);}`) but sadly this isn't allowed either. =(

**Comment by Loirooriol** [comment]
@JJanz Rather than counters, you may have better luck with `inherit()`, already approved by CSSWG resolution: https://github.com/w3c/csswg-drafts/issues/2864#issuecomment-816280875

```css
@property --tag-depth {
  syntax: "<integer>";
  inherits: true;
  initial-value: 0;
}
.indented {
  --tag-depth: calc(1 + inherit(--tag-depth));
  margin-left: calc(1.5rem * var(--tag-depth);
}
```

**Comment by JJanz** [comment]
> Rather than counters, you may have better luck with `inherit()`, already approved by CSSWG resolution: [#2864 (comment)](https://github.com/w3c/csswg-drafts/issues/2864#issuecomment-816280875)

Neat! I couldn't get to it in my research! Thanks!


> Note that with counters aren't probably what you want for depth

Yes, I was aware my example could fail somehow (I did see it fail by a quick test **writing** `counter(tag-depth)` on `content` property but thought maybe there was a way a through - and I was out of options, anyway), as in fact I made it from an actual code I started in the lines of the **second** example I gave. First made it when sure it would work and on a train of thought that is exactly what inherit() seems to do but, on no avail, following research brought me to counters and here.

Again, thank you so much for your help! =)

**Comment by alystair** [comment]
Relates to spec item: https://drafts.csswg.org/css-lists-3/#counter-functions

A major styling issue front-end developers and designers experience is when they are tasked with dynamically changing styling of repeated elements. As of writing this is normally done either via swaths of `:nth-child()` and/or Javascript.

Permitting the use of counters within calc() can greatly simplify and condense such verbose code.

Let's say we wanted to animate each of the following list items in a cascade.

```html
<ol>
	<li>Foo</li>
	<li>Bar</li>
	<li>Bizz</li>
	<li>Buzz</li>
</ol>
```
Currently it has to be manually like so:

```css
@keyframes reveal { 0% { opacity:0; } }
ol>* { animation: reveal both 3s; }
li:nth-child(1) { animation-delay: 0.5s; }
li:nth-child(2) { animation-delay: 1s; }
li:nth-child(3) { animation-delay: 1.5s; }
li:nth-child(4) { animation-delay: 2s; }
```

When the number of items or timing changes, each element must be revised.

However, counter values allow code that is:
- much more succinct, less verbose
- easier to revise and maintain
- less fragile

```css
@keyframes reveal { 0% { opacity:0; } }
ol>* {
	animation: reveal both 3s;
	animation-delay:calc( counter-value(list-item) / 0.5s );
}
```

This opens up the world to so many other effects that were limited to JS and annoying manual composition in a clean form...  although my creative juices are running low - I bet someone more creative such as @argyleink can come up with many other fun uses.

Perhaps limiting use to the special reserved [list-item](https://drafts.csswg.org/css-lists-3/#list-item-counter) to start with could guarantee some sort of baseline to get the rest sorted? My concern as an end user is that other potential methods (Such as the `@property` example shown in a comment above) don't have the clarity that a counter value, 'n' of nth-child, or the proposed `sibling-index()` have...

**Comment by john-hascall** [comment]
I too was sad that this did not work.  I had to turn something elegant:

```
.nav ol li.active ~ .indicator {
  transform: translateX(calc(70px * calc(counter(list-item) - 1)));
}
```

into a whole string of hideous and fragile:


```
.nav ol li:nth-child(0).active ~ .indicator {
  transform: translateX(calc(70px * 0));
}
    ... cases 1..11 here ...

.nav ol li:nth-child(12).active ~ .indicator {
  transform: translateX(calc(70px * 12));
}
```

**Comment by sirisian** [comment]
If you need another example, this can be useful with overlapping grid items. In this example the user wants item1 to span both columns and item2 to be in the second column for each row. Doing this without counters requires explicit row numbers.

```css
#grid {
  display: grid;
  grid-template-columns: repeat(2, 20px);
  counter-reset: row;
}
.item1 {
  counter-increment: row;
  grid-column: 1 / span 2;
  grid-row: counter(row);
}
.item2 {
  grid-column: 2;
  grid-row: counter(row);
}
```
```html
<div id="grid">
  <div class="item1">1</div>
  <div class="item2">1</div>
  <div class="item1">2</div>
  <div class="item2">2</div>
</div>
```

**Comment by mnik01** [comment]
👍 
+1 on this! Very hot

**Comment by LeaVerou** [comment]
> 👍 +1 on this! Very hot

Hi @mnik01 and welcome to the csswg repo! Just a quick note that since these discussions can get quite long, we tend to avoid posting +1s etc unless we also have something else to say in addition to the expression of support and use reactions instead to simply express support.

**Comment by brandonmcconnell** [comment]
Hey all, as the scope of this issue has changed quite a bit from its original concept/spec, it may be worth considering changing the name on the ticket, so it can be found more easily.

I essentially duplicated the more recent evolution of this issue in #8981.

---

@tabatkins made a good point in my related ticket that using `n` to reference the `n` from an `nth-child` statement wouldn't work well since multiple `nth-child` statements can be used on a single selector.

One possibility, introducing a new convention, for this purpose would be to add a second optional argument to `nth-child` for a CSS custom property to assign the iterator value to, like this:

```css
element:nth-child(n + 1, --n) {
  transition: /* ... */;
  --delay: calc(var(--n) * 200ms);
  transition-delay: var(--delay);
}
```

In either case, we should still discuss all the pros/cons/implications of allowing selectors to also be property-declarative in this way, as this would set a very new precedent not before used in CSS afaik.

Another consideration of using a second argument in those pseudo-selector expressions for this is that it would also beg the question, "Would each `:nth-child()` second arg value yield a unique value per its expression, or would this iterator value match the `nth` **matched** element?"

```css
element:nth-child(n, --m):nth-child(n + 1, --n) {
  /* are `m` and `n` equal?
     it's a bit unclear. */
}
```

Alternatively, should an iterator require a completely different syntax altogether, such as the following example, where the custom property name is used completely outside the expression itself, after the selector and before the style block?

```css
element:nth-child(n):nth-child(n + 1) --n {
  transition: /* ... */;
  --delay: calc(var(--n) * 200ms);
  transition-delay: var(--delay);
}
```

**Comment by brandonmcconnell** [comment]
@rthrejheytjyrtj545 That makes sense. In that case, it would probably make the most sense as a `<calc-constant>` that needs to be used in a property (standard or custom) with the styles under a selector with an `:nth-child()` rule, or similar, which would match the `nth` matched element.

**Comment by Loirooriol** [comment]
Counters can be modified within all the subtree in abitrary ways. If you just want the element index and sibling count, refer to #4559 instead.

**Comment by brandonmcconnell** [comment]
@Loirooriol Thanks for pointing me to that issue. Adding my thoughts there. 🙂

**Comment by Loirooriol** [comment]
@tavin `calc(counter(foo, lower-alpha))` makes no sense. Also it's confusing if the same expression can resolve to entirely different types depending on the context.

**Comment by tavin** [comment]
@Loirooriol we can debate whether `counter(foo)` is confusing in a numeric expression, but any idea whether this or even `calc(counter-value(foo))` has a real chance of being adopted?

**Comment by Loirooriol** [comment]
Doesn't seem much likely to happen due to https://github.com/w3c/csswg-drafts/issues/1026#issuecomment-341295647

**Comment by kizu** [comment]
> Doesn't seem much likely to happen due to https://github.com/w3c/csswg-drafts/issues/1026#issuecomment-341295647

Could something akin to setting up a particular scope for a certain counter help with that?

Similar to how we can use `timeline-scope` to set up a scope for an animation? So something like a `counter-value-scope: foo` on a container switching anything that tries to access this particular counter's value inside to this non-parallel/delayed slower mode? This way nothing would be affected by default, and authors couldn't easily “break” the performance of everything, as they would be required to explicitly state which counter should receive this behavior.

**Comment by tomasdev** [comment]
Alternatively maybe some sort of nth-usage inside calc?

```
foo:nth-child(n) {
  animation-delay: calc(n) * 20ms; 
}
```

**Comment by Loirooriol** [comment]
@tomasdev See #8981, "We cannot make values dependent on what kind of selector they were applied with".
However, in this case you could use `animation-delay: calc(sibling-index() * 20ms);` which is already in the [spec](https://drafts.csswg.org/css-values-5/#tree-counting).

**Comment by alystair** [comment]
> @tomasdev See #8981, "We cannot make values dependent on what kind of selector they were applied with". However, in this case you could use `animation-delay: calc(sibling-index() * 20ms);` which is already in the [spec](https://drafts.csswg.org/css-values-5/#tree-counting).

Thanks for the update re: `sibling-index()` @Loirooriol! This would permit me to achieve my intended use showcased in my [2022 comment above](https://github.com/w3c/csswg-drafts/issues/1026#issuecomment-1247631054). Hope to see it implemented in the future. When used along with `sibling-count()` it can generate a sort of seed based randomness can be used artistically (confetti, etc)... although I hope it works on `#foo::before{content:''}` elements!

**Comment by Loirooriol** [comment]
@alystair For randomness, see https://drafts.csswg.org/css-values-5/#randomness

For `sibling-*()` on pseudo-elements, see #9573

**Comment by LeaVerou** [comment]
`sibling-index()` is nice but `counter()` is more powerful, as it doesn't just count siblings (and can be incremented with any increment). I would much rather see something that works like `counter()` but produces a number.

**Comment by petergoes** [comment]
Quite recently we got an upgrade to the `attr()` function where it now can receive a `type()` function.
Could the `counter()` function get a similar upgrade so we could do something like this?

```css
counter(counterName type(<number>))
```

Reading through the comments of this issue, I understand that sometimes you can end up in a loop which is undesirable. 
As an alternative, could the `calc` function get an upgrade so we could do something like this:

```css
animation-delay: calc( counter(item) type(<number>) * 0.5s );
```

Here the returned value of `counter(item)` is converted to a `<number>` by the `calc` function.

Personally, I would prefer the first option.

**Comment by cletusw** [comment]
Chrome intends to ship `sibling-index()` in the next stable version (138)!

https://groups.google.com/a/chromium.org/g/blink-dev/c/opsqx1cBPyc/m/pJaRwh7UAAAJ

**Comment by 5e-Cleric** [comment]
> Chrome intends to ship `sibling-index()` in the next stable version (138)!
> 
> https://groups.google.com/a/chromium.org/g/blink-dev/c/opsqx1cBPyc/m/pJaRwh7UAAAJ


The feature is still some ways of reaching its full potential. Whenever the style queries start processing the values of custom properties, we could have something like this with the new ternary operator `if()`:

```html

<body>
    <div></div>
    <div></div>
    <div></div>
    <div></div>
    <div></div>
    <div></div>
</body>

```

```css

body {

   div {
      --number:sibling-index();
      --min: 5;
      --max: 7;
      --value: clamp(0, min(
                  calc((var(--number) - var(--min)) * 1000),
                  calc((var(--max) - var(--number)) * 1000)
               ), 1);

      background:if(style(--value: 1): blue; else: red);
      
   }

}
```

And you could style ranges of pages "relatively" easily.

This doesn't work because right now style() queries do not compute the custom property, just evaluate the value as a string, so they evaluate the clamp as a string vs the integer as a string.

I hope this is changed in the future.

**Comment by Loirooriol** [comment]
@5e-Cleric This already works in Chromium Canary, just register the property as `<number>`: https://software.hixie.ch/utilities/js/live-dom-viewer/saved/13828

**Comment by 5e-Cleric** [comment]
Damn, spent 2h yesterday figuring this out and trying different combinations, i was using counters before i found sibling-index() and custom properties were messing assignment somehow, so i figured they would not allow for the interpretation!

Thanks for that, then, it is indeed a great addition with lots of potential.

(This example in particular was thought because i create documents with  markdown+css, and i wanted an easy programatic way to get sections styled)

**Comment by Rezi** [comment]
I made a hacky solution to get counter value into a variable. Here is an explaining article (including codepen) https://dev.to/rezi/css-counting-magic-converting-counter-values-to-variables-e9m
It would be great if this can be done in an easy way.

**Comment by dragontheory** [comment]
## Smell test

Just had this conversation . . . 

Dev: "CSS is broken." 
Me: "Yeah, funny meme."
Dev: "No, seriously. LIs don't even work. Numbers show, but that's it. Can't do [expletive] with them. Same with counters. Half baked."
Me: "It's purposeful".
Dev: ". . . F'ing hate CSS." 

## Fail

The browser already computes counter values for every list item. To render `::marker`. To fill `counter()` in `content:`. To honor `<ol start>`, `<ol reversed>`, `<li value>`, `@counter-style`, scoped `counter-reset`. The user sees the values on screen.

**_CSS can't read them._**

That's the whole problem. The engine knows. The language refuses to surface what the engine knows.

```css
li { --i: counter-value(list-item); }
```
Returns `<integer>`. Usable in `calc()`, custom properties, `@container style()`. Same shape as `sibling-index()`, but counter-aware.

[@tabatkins endorsed this exact shape](https://github.com/w3c/csswg-drafts/issues/1026#issuecomment-337699774): typed retrieval, parallel to the `attr()` `type()` extension. [@LeaVerou said it directly](https://github.com/w3c/csswg-drafts/issues/1026#issuecomment-2471994369): *"I would much rather see something that works like `counter()` but produces a number."*

> "We have `sibling-index()`"

`sibling-index()` ignores `<ol start>`, `<ol reversed>`, `<li value>`, custom `@counter-style`, nested counter scopes, and author-defined increments. It answers a different question.

Authors who use semantic HTML correctly, declaring `start`, `value`, `reversed`, get their semantics honored visually, but _**not programmatically**_. The browser respects what they declared, but _**CSS doesn't**_.

If `sibling-index()` were the answer, the [Chromium `@property` hack](https://github.com/w3c/csswg-drafts/issues/1026#issuecomment-1029205510) wouldn't exist. Authors chose the hack anyway, because it does what `sibling-index()` can't.

> This would break parallel style computation. [@upsuper, 2017](https://github.com/w3c/csswg-drafts/issues/1026#issuecomment-341435007): 

1. **`sibling-index()` shipped.** Same structural shape, tree-derived integer at style time. Engines parallelized what they could and serialized the rest.
2. **`:has()` shipped.** Twenty years of *"this breaks invalidation"* turned out to be bounded. Engines built new strategies.
3. **The Chromium `@property` hack works.** Register a custom property as `<number>`; counter values flow through. If parallelization were broken, neither would the hack.

Already answered. [@kizu's `counter-value-scope`](https://github.com/w3c/csswg-drafts/issues/1026#issuecomment-1633107722) covers any remaining concern: opt-in to the slower path per counter. Authors who don't use it pay nothing.

Paged-media counters (`page`, `pages`, `target-counter()`). [@faceless2's analysis](https://github.com/w3c/csswg-drafts/issues/1026#issuecomment-669416783) have real layout-feedback loops. Defer them. Different problem.

Document-flow counters don't have that shape. Tree + counter rules, resolved before layout. Same dependency direction as `sibling-index()`.

> "Few authors need this"

This dismissal has been the response for nine years. Yet:

- Demand (frustration) keeps accumulating.
- Workarounds have multiplied: `:nth-child()` chains, JavaScript walkers, `data-*` mirrors, the `@property` hack.
- Implementations have demonstrated the capability.
- And CSS advocates like myself have little to justify it.

An author community writing blog posts for an undocumented hack is not a niche. The question is not whether authors need this. The question is how much longer the standard will require us to prove it.

**Comment by Crissov** [comment]
See #12488:

> So `parse(number counter(...))` would indeed convert a counter value into an integer.

**Comment by Loirooriol** [comment]
@dragontheory All your links just point to this issue, would be useful to link to the specific comments you are addressing.

**Comment by dragontheory** [comment]
> [@dragontheory](https://github.com/dragontheory) All your links just point to this issue, would be useful to link to the specific comments you are addressing.

Thank you for reading my comment. Links fixed. 

Now that there is some distance from my conversation with the fellow dev, I see that my comment here could have been less emotional and reactionary. I apologize. Thank you all for doing what you do and enabling native modern CSS to be what it is today. CSS is truly awesome!

**Comment by Gonzalo-D-Sales** [comment]
Hello! I apologize if this is not 100% related, but I have found a genuine use case for enabling `counter()` to work inside of `if()` functions. 

I develop themes for an Electron application (Obsidian) where the DOM is generated by JavaScript and cannot be modified. There are containers that visually have no children (all children are `display: none`) but are not structurally empty, since they contain a wrapper element that itself has no visible descendants. The only way to conditionally style the container right now is by using `:has()`, which is costly and already frequently used in the application's own stylesheet. A counter-based approach allows me to detect "visually empty" containers without using `:has()`.

I observed that `counter-increment` does not increment on an element with `display: none`. This means a counter scoped to a parent and incremented by its children effectively counts only **visible** children, which `sibling-count()` does not since it counts all DOM siblings regardless of their display state. Counters correctly track this information, and they render accurately when called in `content`. But there is currently no way to access their result outside of `content`, or crucially, to use the result to branch via `if()` or `@container` style queries.

Now as for implementing such a change, I believe there is a good precedent for it. From my understanding, `sibling-count()` already returns an integer value for use in `calc()` and conditional statements. Either a hypothetical `counter-value()` function, or allowing `counter()` to resolve to an `<integer>` when used outside of `content`, would enable visible-child-aware styling that is intuitive and performant compared to using `:has()`.

This is what I have attempted so far:

1. `counter-reset: children` on a parent container
2. `counter-increment: children` on each child element
3. Storing the result in a variable `--count: counter(children)`
4. Styling the parent based on this result `if(style(--count: 0): /* hide parent */)`

The first three work, but as expected, I run into a roadblock at the last point. The variable stores the **unresolved** `counter()` tokens, not the resolved value. From my limited understanding, the only context where `counter()` currently resolves to a usable value is when used in `content` (where it resolves to a string). Therefore, my `if()` styling does not work.

There is an additional complication, that being that the updated counter value only shows up on an `::after` pseudo-element of the parent and not the parent itself. This means that I cannot access the updated `counter()` value from the parent element I actually want to style, only from its pseudo-element which appears *after* all the children element in the DOM. 

What this means is that the only options for conditionally styling the parent, even if I am able to access `counter()`'s value outside of `content`, is to use the pseudo-element to hide or cover the parent, or by grabbing the pseudo-element's `content` using a hypothetical `content(after)` statement. This has been considered already in the past: "The CSS generated content module introduces six yet-to-be implemented CSS functions including `content()`, `string()`, and `leader()`, and the three [`[<target>](https://developer.mozilla.org/en-US/docs/Web/CSS/Reference/Properties/content#target)`](https://developer.mozilla.org/en-US/docs/Web/CSS/Reference/Properties/content#target) functions `target-counter()`, `target-counters()`, and `target-text()`." in https://developer.mozilla.org/en-US/docs/Web/CSS/Guides/Generated_content#functions.

On one last note, I have tried to read and understand the rationale against implementing these changes, due to the risk of circular dependencies. However, I believe the use case I suggested is simple and isolated enough to mostly sidestep the issue. After all, there is no use of `calc()` in my use case, and all of the child elements are already `display: none` so the counter reads 0 on the first layout pass. Hiding the parent element (and thus removing all its children from the rendering tree) based on this result does not change the counter, since the counter already resolved to 0 to begin with.

**Comment by Loirooriol** [comment]
@Gonzalo-D-Sales What you are proposing is inherently circular in general.
